platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/ExceptionProblemRegistry.java

```diff
@@ -35,14 +40,119 @@
 @ServiceScope(Scope.BuildSession.class)
 public class ExceptionProblemRegistry {
 
-    private final Multimap<Throwable, Problem> problemsForThrowables = Multimaps.synchronizedMultimap(HashMultimap.<Throwable, Problem>create());
-
+    private final Multimap<Throwable, Problem> problemsForThrowables = Multimaps.synchronizedMultimap(MultimapBuilder.linkedHashKeys().linkedHashSetValues().<Throwable, Problem>build());
 
     public void onProblem(Throwable exception, Problem problem) {
         problemsForThrowables.put(exception, problem);
     }
 
-    public Map<Throwable, Collection<Problem>> getProblemsForThrowables() {
-        return problemsForThrowables.asMap();
+    public ProblemLookup getProblemLookup() {
+        return new DefaultProblemLookup();
```

Comment:
❓ Why do we recreate this expensive class on each call instead of keeping the last one around?

Reply:
It raises concurrency issues if we cache the problem lookup. At the same time, we can make the expensive part (ie. `initLookup()` lazy).